mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -43,6 +55,379 @@ struct CompactBlockRow {
     data: Vec<u8>,
 }
 
+#[async_trait]
+pub trait ZRpcOps {
+    async fn get_block_height(&mut self) -> Result<u64, MmError<UpdateBlocksCacheErr>>;
+
+    async fn scan_blocks(
+        &mut self,
+        start_block: u64,
+        last_block: u64,
+        on_block: &mut (dyn FnMut(TonicCompactBlock) -> Result<(), MmError<UpdateBlocksCacheErr>> + Send + Sync),
+    ) -> Result<(), MmError<UpdateBlocksCacheErr>>;
+
+    async fn check_watch_for_tx(&mut self, tx_id: TxId, attempts: &mut i32, watch_for_tx: &mut Option<TxId>);
+}
+
+#[async_trait]
+impl ZRpcOps for CompactTxStreamerClient<Channel> {
+    async fn get_block_height(&mut self) -> Result<u64, MmError<UpdateBlocksCacheErr>> {
+        let request = tonic::Request::new(ChainSpec {});
+        let block = self.get_latest_block(request).await?;
+        let res = block.into_inner().height;
+        Ok(res)
+    }
+
+    async fn scan_blocks(
+        &mut self,
+        start_block: u64,
+        last_block: u64,
+        on_block: &mut (dyn FnMut(TonicCompactBlock) -> Result<(), MmError<UpdateBlocksCacheErr>> + Send + Sync),
+    ) -> Result<(), MmError<UpdateBlocksCacheErr>> {
+        let request = tonic::Request::new(BlockRange {
+            start: Some(BlockId {
+                height: start_block,
+                hash: Vec::new(),
+            }),
+            end: Some(BlockId {
+                height: last_block,
+                hash: Vec::new(),
+            }),
+        });
+        let mut response = self.get_block_range(request).await?;
+        while let Some(block) = response.get_mut().message().await? {
+            debug!("Got block {:?}", block);
+            on_block(block)?;
+        }
+        Ok(())
+    }
+
+    async fn check_watch_for_tx(&mut self, tx_id: TxId, attempts: &mut i32, watch_for_tx: &mut Option<TxId>) {
```

Comment:
It's not required and even a bit confusing to pass mutable references to `attempts` and `watch_for_tx` here. `attempts` must be fully controlled by this method itself. `watch_for_tx` must be controlled by the code that calls `check_watch_for_tx`.

Reply:
done